### PR TITLE
[write] add `params` argument to `wb_add_data_table()`

### PR DESCRIPTION
**Pull request by JanMarvin**
Which allows to filter a table with `choose`. Similar to how it is done in `wb_add_pivot_table()`.

``` r
library(openxlsx2)

choose <- c(cyl = "x %in% c(4, 8)", am = c("x != 1"))
  
# create workbook
wb <- wb_workbook() %>%
  wb_add_worksheet() %>%
  wb_add_data_table(x = mtcars, params = list(choose = choose))
wb_to_df(wb, skip_hidden_rows = TRUE)
#>     mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> 6  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
#> 8  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
#> 9  24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2
#> 10 22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2
#> 13 16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3
#> 14 17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3
#> 15 15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3
#> 16 10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4
#> 17 10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4
#> 18 14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
#> 22 21.5   4 120.1  97 3.70 2.465 20.01  1  0    3    1
#> 23 15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2
#> 24 15.2   8 304.0 150 3.15 3.435 17.30  0  0    3    2
#> 25 13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4
#> 26 19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2

# write_xlsx
wb <- write_xlsx(x = mtcars, as_table = TRUE, params = list(choose = choose))
wb_to_df(wb, skip_hidden_rows = TRUE)
#>     mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> 6  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
#> 8  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
#> 9  24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2
#> 10 22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2
#> 13 16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3
#> 14 17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3
#> 15 15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3
#> 16 10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4
#> 17 10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4
#> 18 14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
#> 22 21.5   4 120.1  97 3.70 2.465 20.01  1  0    3    1
#> 23 15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2
#> 24 15.2   8 304.0 150 3.15 3.435 17.30  0  0    3    2
#> 25 13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4
#> 26 19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2
```

### CODE REVIEW

**Comment by JanMarvin** [comment]
This PR requires testing. Especially how it works in comparison with spreadsheet software. How is does this work compared to spreadsheet software and how does this work in comparison with other strange things thrown at it? Like custom filters.

File: R/class-workbook.R [hunk 1]
```diff
@@ -3431,19 +3434,6 @@ wbWorkbook <- R6::R6Class(
...
-      autofilter <- if (withFilter) {
-        if (!isFALSE(totalsRowCount)) {
-          # exclude total row from filter
-          rowcol         <- dims_to_rowcol(ref)
-          autofilter_ref <- rowcol_to_dims(as.integer(rowcol[[2]])[-length(rowcol[[2]])], rowcol[[1]])
-        } else {
-          autofilter_ref <- ref
-        }
-        xml_node_create(xml_name = "autoFilter", xml_attributes = c(ref = autofilter_ref))
-      }
```
Comment (JanMarvin) [inline R/class-workbook.R:3445]: what is autofilter if not TRUE? NULL? Something else?

**Comment by JanMarvin** [comment]
Maybe it is better to not use params but choose directly, but I'm not gonna change it right now

**Comment by JanMarvin** [comment]
In openxml it is possible to have at least two conditions per column filter. This is not covered by this PR and wont be included.

This

```R
choose <- c(
  cyl = "x > 14 & x <= 20",
  am  = "x == 3 | x == 5"
)
```

should result in this

```XML
 <autoFilter ref="A1:K33">
  <filterColumn colId="0">
   <customFilters and="1">
    <customFilter operator="greaterThan" val="14" />
    <customFilter operator="lessThanOrEqual" val="20" />
   </customFilters>
  </filterColumn>
  <filterColumn colId="9">
   <filters>
    <filter val="3" />
    <filter val="5" />
   </filters>
  </filterColumn>
 </autoFilter>
```

Conditions with `&` are connected with `and="1"`. Probably custom filter needs an operator "equal" for cases like `x == 0 | x > 1`. This is not implemented and would require improvements to `create_conditions()` and to `reverse_conditions()`.

**Comment by klin333** [comment]
yep gave it a try. worked very nicely. would love for this feature to be merged into master!

**Comment by JanMarvin** [comment]
Hi @klin333 , thanks for reaching out! Though I had two or three things in mind that I want feedback for.

How does `openxlsx2` behave compared to Excel:
1.  if two tables are next to each other and have filters for "disjoint quantitys"? E.g., what happens if I have two tables next to each other, lets say `mtcars` and one table has a filter on `am == 0` and the other has a filter `am == 1`.
```R
wb <- wb_workbook()$add_worksheet()$
  add_data_table(dims = wb_dims(x = mtcars), x = mtcars)$
  add_data_table(dims = wb_dims(x = mtcars, from_dims = wb_dims(x = mtcars), right = 2), x = mtcars)
```
2. if the table is placed on a spreadsheet with a filter on the worksheet level. E.g. there is a table below a `mtcars`object written directly into the spreadsheet. Will this filter impact the table? Something like:
```R
wb <- wb_workbook()$add_worksheet()$
  add_data(dims = wb_dims(x = mtcars), x = mtcars)$
  add_data_table(dims = wb_dims(x = mtcars, from_dims = wb_dims(x = mtcars), below = 2), x = mtcars)$
  add_filter(rows = 1, cols = seq_along(mtcars))
```
3. If a row is hidden, but the user requests an unhide for this specific row. Does the filter have to be adjusted?

Basically: We can add filters in Excel in filters in `openxlsx2`, but are many filters allowed? Do they impact each other? The filter simply hides rows.
Finally, isn't the `params` argument a little to confusing? Is there a better, more consistent way?

**Comment by JanMarvin** [comment]
Thanks again, @lkoribsky. I appreciate the detailed testing. Constructing and running these tests is time-consuming, and since I wrote this code a while ago, your feedback is very welcome.

My main concern is preventing 'unpleasant surprises' for the end user. Especially those users that are not all that familiar with spreadsheet software might not be aware how spreadsheets behave. That the sum of visible rows can differ significantly from the sum of all rows in a sheet. Your tests have at least reassured me that the logic is performing as intended.

I chose the name `params` to stay consistent with pivot tables, where its possible to 'squash' various arguments into one place. This gives us the flexibility to eventually include other features, like the number formats from our previous discussion. But I wrote this PR back in September 2024, so I need to look into it myself again. Ideally, I’d like to add 'AND' and 'OR' conditions as well.

However, taking 'baby steps' is likely the best path forward. We can always refine and improve the functionality. And I would like to avoid experimenting with the API. Every change is potentially breaking and I really dislike it when others do that to me :)

**Comment by lkoribsky** [comment]
No worries @JanMarvin  
As it was mentioned in another thread, we are using this package for generating Excel files base on the specification in yaml file, so we can try to test more features which you develop to assure you, it is working as spreadsheet software. This probably makes also our development easier
